src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs

```diff
@@ -0,0 +1,228 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.IO;
+using System.Linq;
+using System.Threading;
+
+using Microsoft.VisualStudio.LanguageServices.ProjectSystem;
+using Microsoft.VisualStudio.ProjectSystem.Logging;
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+
+namespace Microsoft.VisualStudio.ProjectSystem.LanguageServices
+{
+    /// <summary>
+    ///     Applies <see cref="IProjectVersionedValue{T}"/> values to a <see cref="IWorkspaceProjectContext"/>.
+    /// </summary>
+    [Export(typeof(IApplyChangesToWorkspaceContext))]
+    internal class ApplyChangesToWorkspaceContext : OnceInitializedOnceDisposed, IApplyChangesToWorkspaceContext
+    {
+        private const string DesignTimeRuleName = CompilerCommandLineArgs.SchemaName;
+        private readonly ConfiguredProject _project;
+        private readonly ICommandLineParserService _commandLineParser;
+        private readonly IProjectLogger _logger;
+
+        private IWorkspaceProjectContext _context;
+        private ExportLifetimeContext<IWorkspaceContextHandler>[] _handlers;
+
+        [ImportingConstructor]
+        public ApplyChangesToWorkspaceContext(ConfiguredProject project, ICommandLineParserService commandLineParser, IProjectLogger logger)
+        {
+            _project = project;
+            _commandLineParser = commandLineParser;
+            _logger = logger;
+
+            WorkspaceContextHandlerFactories = new OrderPrecedenceExportFactoryCollection<IWorkspaceContextHandler>();
+        }
+
+        [ImportMany]
```

Comment:
can this be imported in the constructor instead?

Reply:
Usually we don't it for OrderPrecence because the results can change under the covers based on capabilities. In this case, however, we're not expecting them to change, so I agree and moved to constructor.